From c1229340f1380149f44cbe48854afe507971aa18 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Mon, 24 Feb 2025 17:38:44 +0800
Subject: [PATCH] hw/xfree86: workaround bad firmware boot VGA reporting on
 MIPS Loongson devices

Loongson (MIPS) platforms based on (at least) the 7A bridge chip has
faulty firmware that does not export the correct boot VGA device to the
operating system, causing the X server to mistake the non-connected output
as the primary match for X output - which then causes the X server to fail
to probe for the right driver, in turn causing X server to fail at
startup.

From our user report, this motherboard/firmware combination is known to
have this issue (there is probably more):

  BM6J74 (Kunlun 4.00.0024, OEM version LJ74B00BPS)

The original patch from Loongson/UOS uses an extra call to
`xf86platformPrimary()' to override the result from the above call to
`xf86platformProbe()', prompting X to use the first platform device found
on the device (the first video card/GPU probed by the kernel (for
modesetting or simply as firmware output).

This will most certainly break other platforms without this issue, so
limit this workaround to little-endian MIPS64 with N64 ABI for now (for
which AOSC OS only supports Loongson 3 devices).

[Where this patch was first shown to the public...]
Link: https://github.com/deepin-community/xorg-server/pull/11

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 hw/xfree86/common/xf86Bus.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

Index: xorg-server/hw/xfree86/common/xf86Bus.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Bus.c
+++ xorg-server/hw/xfree86/common/xf86Bus.c
@@ -241,6 +241,34 @@ xf86BusProbe(void)
 {
 #ifdef XSERVER_PLATFORM_BUS
     xf86platformProbe();
+/*
+ * Loongson (MIPS) platforms based on (at least) the 7A bridge chip has
+ * faulty firmware that does not export the correct boot VGA device to the
+ * operating system, causing the X server to mistake the non-connected
+ * output as the primary match for X output - which then causes the X
+ * server to fail to probe for the right driver, in turn causing X server
+ * to fail at startup.
+ *
+ * From our user report, this motherboard/firmware combination is known
+ * to have this issue (there is probably more):
+ *
+ *   BM6J74 (Kunlun 4.00.0024, OEM version LJ74B00BPS)
+ *
+ * The original patch from Loongson/UOS uses an extra call to
+ * `xf86platformPrimary()' to override the result from the above call to
+ * `xf86platformProbe()', prompting X to use the first platform device
+ * found on the device (the first video card/GPU probed by the kernel (for
+ * modesetting or simply as firmware output).
+ *
+ * This will most certainly break other platforms without this issue, so
+ * limit this workaround to little-endian MIPS64 with N64 ABI for now for
+ * which AOSC OS only supports Loongson 3 devices).
+ */
+#if (defined(__mips) && defined (_ABI64) && \
+     defined(__BYTE_ORDER__) && defined(__ORDER_LITTLE_ENDIAN__) && \
+     (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__))
+    xf86platformPrimary();
+#endif
     if (ServerIsNotSeat0() && xf86_num_platform_devices > 0)
         return;
 #endif
